src/impl_owned_array.rs

```diff
@@ -59,3 +69,558 @@ where
+        sort_axes_in_default_order(&mut self_);
+
+        unsafe {
+            let array_memory_head_ptr = self_.ptr.as_ptr();
+            let data_end_ptr = data_ptr.add(data_len);
+            debug_assert!(data_ptr <= array_memory_head_ptr);
+            debug_assert!(array_memory_head_ptr <= data_end_ptr);
+
+            // iter is a raw pointer iterator traversing self_ in its standard order
+            let mut iter = Baseiter::new(self_.ptr.as_ptr(), self_.dim, self_.strides);
+            let mut dropped_elements = 0;
+
+            // The idea is simply this: the iterator will yield the elements of self_ in
+            // increasing address order.
+            //
+            // The pointers produced by the iterator are those that we *do not* touch.
+            // The pointers *not mentioned* by the iterator are those we have to drop.
+            //
+            // We have to drop elements in the range from `data_ptr` until (not including)
+            // `data_end_ptr`, except those that are produced by `iter`.
+            let mut last_ptr = data_ptr;
+
+            while let Some(elem_ptr) = iter.next() {
+                // The interval from last_ptr up until (not including) elem_ptr
+                // should now be dropped. This interval may be empty, then we just skip this loop.
+                while last_ptr != elem_ptr {
+                    debug_assert!(last_ptr < data_end_ptr);
+                    std::ptr::drop_in_place(last_ptr as *mut A);
+                    last_ptr = last_ptr.add(1);
+                    dropped_elements += 1;
+                }
+                // Next interval will continue one past the current element
+                last_ptr = elem_ptr.add(1);
+            }
+
+            while last_ptr < data_end_ptr {
+                std::ptr::drop_in_place(last_ptr as *mut A);
+                last_ptr = last_ptr.add(1);
+                dropped_elements += 1;
+            }
+
+            assert_eq!(data_len, dropped_elements + self_len,
+                       "Internal error: inconsistency in move_into");
+        }
+        self.data
+    }
+
+    /// Create an empty array with an all-zeros shape
+    ///
+    /// ***Panics*** if D is zero-dimensional, because it can't be empty
+    pub(crate) fn empty() -> Array<A, D> {
+        assert_ne!(D::NDIM, Some(0));
+        let ndim = D::NDIM.unwrap_or(1);
+        Array::from_shape_simple_fn(D::zeros(ndim), || unreachable!())
+    }
+
+    /// Create new_array with the right layout for appending to `growing_axis`
+    #[inline(never)]
+    fn change_to_contig_append_layout(&mut self, growing_axis: Axis) {
+        let ndim = self.ndim();
+        let mut dim = self.raw_dim();
+
+        // The array will be created with 0 (C) or ndim-1 (F) as the biggest stride
+        // axis. Rearrange the shape so that `growing_axis` is the biggest stride axis
+        // afterwards.
+        let prefer_f_layout = growing_axis == Axis(ndim - 1);
```

Comment:
Is there a reason for preferring the standard layout instead of being based on the current layout?

Reply:
> Is there a reason for preferring the standard layout instead of being based on the current layout?

I'm not sure what you mean. With a swap instead of a rotate, the layout still isn't based on the current layout; it's a modification of the standard layout. The reason for preferring a rotate over a swap is that it's fairly likely that the array being appended to `self` will be in standard layout and have length 1 for the "growing axis".

To consider an example, let's define the "axis order" as the axis indices in order of decreasing absolute stride. For example, a 4-D standard layout array would have axis order

```
[0, 1, 2, 3]
```

Now, let's say we want to append to a 4-D array, and we're calling `change_to_contig_append_layout` to create a new owned array with axis 2 as the largest-stride axis.

- If we do a swap, then the strides of the owned array will correspond to the following axis order:

   ```
   [2, 1, 0, 3]
   ```

   so if we append a standard-layout array of shape e.g. `[5, 5, 1, 5]` to the owned array, the order of iteration of the array being appended to the owned array will not be in increasing memory order.

- If we do a rotate, then the strides of the owned array will correspond to the following axis order:

   ```
   [1, 2, 0, 3]
   ```

   so if we append a standard-layout array of shape e.g. `[5, 5, 1, 5]`, the order of iteration will be in increasing memory order for both the owned array and the array being appended.

Reply:
Thanks, now I totally see what you mean. That gives a good chance that the `Zip` will use a contiguous order.